src/ripple/app/tx/impl/AMMVote.cpp

```diff
@@ -0,0 +1,214 @@
+    // Find the entry with the least tokens and whether the account
+    // has the vote entry.
+    for (auto const& entry : ammSle->getFieldArray(sfVoteSlots))
+    {
+        auto const account = entry[sfAccount];
+        auto lpTokens = ammLPHolds(sb, *ammSle, account, ctx_.journal);
+        if (lpTokens == beast::zero)
+        {
+            JLOG(j_.debug())
+                << "AMMVote::applyVote, account " << account << " is not LP";
+            continue;
+        }
+        auto feeVal = entry[sfTradingFee];
+        STObject newEntry{sfVoteEntry};
+        // The account already has the vote entry.
+        if (account == account_)
+        {
+            lpTokens = lpTokensNew;
+            feeVal = feeNew;
+            foundAccount = true;
+        }
+        // Keep running numerator/denominator to calculate the updated fee.
+        num += feeVal * lpTokens;
+        den += lpTokens;
+        newEntry.setAccountID(sfAccount, account);
+        newEntry.setFieldU16(sfTradingFee, feeVal);
+        newEntry.setFieldU32(
+            sfVoteWeight,
+            static_cast<std::int64_t>(
+                Number(lpTokens) * 100000 / lptAMMBalance));
+
+        // Find an entry with the least tokens/fee. Make the order deterministic
+        // if the tokens/fees are equal.
+        if (!minTokens ||
+            (lpTokens < *minTokens ||
+             (lpTokens == *minTokens &&
+              (feeVal < minFee || (feeVal == minFee && account < minAccount)))))
+        {
+            minTokens = lpTokens;
+            minPos = updatedVoteSlots.size();
+            minAccount = account;
+            minFee = feeVal;
+        }
+        updatedVoteSlots.push_back(std::move(newEntry));
+    }
+
+    // The account doesn't have the vote entry.
+    if (!foundAccount)
+    {
+        auto update = [&]() {
+            STObject newEntry{sfVoteEntry};
+            newEntry.setFieldU16(sfTradingFee, feeNew);
+            newEntry.setFieldU32(
+                sfVoteWeight,
+                static_cast<std::int64_t>(
+                    Number(lpTokensNew) * 100000 / lptAMMBalance));
+            newEntry.setAccountID(sfAccount, account_);
+            num += feeNew * lpTokensNew;
+            den += lpTokensNew;
+            updatedVoteSlots.push_back(std::move(newEntry));
+        };
+        // Add new entry if the number of the vote entries
+        // is less than Max.
+        constexpr std::uint8_t MaxVoteSlots = 8;
+        if (updatedVoteSlots.size() < MaxVoteSlots)
+            update();
+        // Add the entry if the account has more tokens than
+        // the least token holder.
+        else if (lpTokensNew > *minTokens)
+        {
+            auto const entry = updatedVoteSlots.begin() + minPos;
+            // Remove the least token vote entry.
+            num -= Number((*entry)[sfTradingFee]) * *minTokens;
+            den -= *minTokens;
+            updatedVoteSlots.erase(updatedVoteSlots.begin() + minPos);
```

Comment:
Minor. Instead of erasing an element and pushing back a new element, you could move the new element into the to-be-erased element.